source/Timer.cpp

```diff
@@ -0,0 +1,346 @@
+	result.resetFired = resetFired;
+	result.idleMaxSpeed = idleMaxSpeed;
+	result.requirePeaceful = requirePeaceful;
+
+	string reason;
+	auto ait = actions.begin();
+	for( ; ait != actions.end(); ++ait)
+	{
+		reason = ait->second.Validate();
+		if(!reason.empty())
+			break;
+	}
+	if(ait != actions.end())
+	{
+		Logger::LogError("Instantiation Error: Timer action \"" + TriggerToText(ait->first)
+			+ "\" uses invalid " + std::move(reason));
+		return result;
+	}
+	for(const auto &it : actions)
+		result.actions[it.first] = it.second.Instantiate(subs, origin, jumps, payload);
+
+	result.waitTime = waitTime;
+	if(randomWaitTime > 1)
+		result.waitTime += Random::Int(randomWaitTime);
+
+	return result;
+}
+
+
+
+bool Timer::IsComplete() const
+{
+	return isComplete;
+}
+
+
+
+bool Timer::IsOptional() const
+{
+	return optional;
+}
+
+
+
+void Timer::ResetOn(ResetCondition cond, PlayerInfo &player, UI *ui, const Mission &mission)
+{
+	bool reset = cond == resetCondition;
+	reset |= (cond == Timer::ResetCondition::LEAVE_ZONE && resetCondition == Timer::ResetCondition::PAUSE);
+	reset |= (cond == Timer::ResetCondition::LEAVE_SYSTEM && (resetCondition == Timer::ResetCondition::PAUSE
+				|| resetCondition == Timer::ResetCondition::LEAVE_ZONE));
+	if(isActive && reset)
+	{
+		timeElapsed = 0;
+		if(repeatReset || !resetFired)
+		{
+			auto it = actions.find(TimerTrigger::RESET);
+			if(it != actions.end())
+				it->second.Do(player, ui, &mission);
+			resetFired = true;
+		}
+		isActive = false;
+	}
+}
+
+
+
+void Timer::Step(PlayerInfo &player, UI *ui, const Mission &mission)
+{
+	if(isComplete)
+		return;
+	const Ship *flagship = player.Flagship();
+	if(!flagship)
+		return;
+	if((system && flagship->GetSystem() != system) ||
+		(!systems.IsEmpty() && !systems.Matches(flagship->GetSystem())))
+	{
+		ResetOn(Timer::ResetCondition::LEAVE_SYSTEM, player, ui, mission);
+		return;
+	}
+	if(requireIdle || requirePeaceful)
+	{
+		bool shipIdle = true;
+		if(requireIdle)
+			shipIdle = (!flagship->IsThrusting() && !flagship->IsSteering()
+						&& !flagship->IsReversing() && flagship->Velocity().LengthSquared() < idleMaxSpeed);
+		if(requirePeaceful)
+			for(const Hardpoint &weapon : flagship->Weapons())
+				shipIdle &= !weapon.WasFiring();
+		if(!shipIdle)
+		{
+			ResetOn(Timer::ResetCondition::PAUSE, player, ui, mission);
+			return;
+		}
+	}
+	if(requireUncloaked)
+	{
+		double cloak = flagship->Cloaking();
+		if(cloak != 0.)
+		{
+			ResetOn(Timer::ResetCondition::PAUSE, player, ui, mission);
+			return;
+		}
+	}
+	if(proximity > 0.)
+	{
+		bool inProximity = false;
+		if(proximityCenter || !proximityCenters.IsEmpty())
+		{
+			for(const StellarObject &proximityObject : system->Objects())
+				if(proximityObject.HasValidPlanet() && (proximityCenter == proximityObject.GetPlanet() ||
+					proximityCenters.Matches(proximityObject.GetPlanet())))
+				{
+					double dist = flagship->Position().Distance(proximityObject.Position());
```

Comment:
I think we should cache the proximityobject

Reply:
Do you mean we should build a cache of all the specified proximityObjects?

I think that should be possible, but once again I'm having trouble with type modifiers—if I add a `std::list<StellarObject> proximityCache` to the Timer definition, and attempt to populate it in the `Instantiate()` method by simply copying the above loop in, I can't seem to get it to allow me to call `proximityCache.push_back(proximityObject)`, even if I try changing the list to hold `const`s...

Reply:
Did you try a pointer?

Reply:
Which kind of pointer? Try it how? I did try to figure out if it could be done with a `shared_ptr` (with little success), but I'm still pretty new at handling all the different kinds of pointers and references that modern C++ supports. (I'd also be happy to discuss it further on Discord when & if that's convenient.)

Reply:
I was thinking about a const StellarObject*

Reply:
Also a std::set might be the right thing for the job?

Reply:
With the cache declared as `std::list<const StellarObject *> proximityCache;`, and the loop as

```
	for(const StellarObject &proximityObject : system->Objects())
		proximityCache.push_back(proximityObject);
```
(leaving out the lengthy conditional, which is presumably irrelevant for this problem), I get this build error

```
endless-sky/source/Timer.cpp:242:19: error: no matching member function for call to 'push_back'
                        proximityCache.push_back(proximityObject);
                        ~~~~~~~~~~~~~~~^~~~~~~~~
/usr/include/c++/v1/list:986:10: note: candidate function not viable: 'this' argument has type 'const std::list<const StellarObject *>', but method is not marked const
    void push_back(value_type&& __x);
         ^
/usr/include/c++/v1/list:1011:10: note: candidate function not viable: 'this' argument has type 'const std::list<const StellarObject *>', but method is not marked const
    void push_back(const value_type& __x);
         ^
1 error generated.
```

(cleaned up slightly for readability, but not removing any of the salient details)

Reply:
You are trying to write data in a const marked function?

Reply:
Ah! You're right; thanks for catching that dumb mistake. Correcting it to `result.proximityCache.push_back(&proximityObject)` appears (provisionally) to work.

I will see about adjusting the rest of the code to use that cache.